[DRAFT] Upgrade to 2.7.0-rc3

gaetbout opened the pull request:
<!--- Please provide a general summary of your changes in the title above -->

## Pull Request type

<!-- Please try to limit your pull request to one type; submit multiple pull requests if needed. -->

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [x] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [x] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):


## What is the new behavior?

There might be some changes in the API. Should we also deprecate some stuff or can we just force to upgrade by removing them?

gaetbout commented:
@0xLucqs Ready for review. 
An upgrade to latest edition should still be performed, will open an issue

0xLucqs commented:
flemme de review je close bonne journée

0xLucqs commented on packages/ascii/src/tests/test_ascii_integer.cairo:
```diff
@@ -1,99 +1,99 @@
-    assert_eq!(*ascii.at(0), '1157920892373161954235709850086', "max u256 wrong first felt");
-    assert_eq!(*ascii.at(1), '8790785326998466564056403945758', "max u256 wrong second felt");
-    assert_eq!(*ascii.at(2), '4007913129639935', "max u256 wrong third felt");
+    assert_eq!(ascii.len(), 3);
```
why ?

0xLucqs commented on packages/data_structures/src/byte_reader.cairo:
```diff
@@ -562,19 +576,10 @@ impl ByteReaderLenImpl<T, +Len<T>> of Len<ByteReaderState<T>> {
-impl ByteReaderIndexViewImpl<
-    T, impl TIndexView: IndexView<T, usize, @u8>
-> of IndexView<ByteReaderState<T>, usize, @u8> {
+impl IntoU8Impl of Into<u8, @u8> {
```
what am i watching here ? is this needed for some trait ?

gaetbout replied:
this is the one thing I'm a bit puzzled, and had a lot pf back and forth with Ariel on Discord.
This is the only way I found to have this file working.
If you'd be kind enough to look into it, that'd be nice...

0xLucqs commented on packages/data_structures/src/byte_reader.cairo:
```diff
@@ -161,7 +171,10 @@ pub trait ByteReader<T> {
-impl ByteReaderImpl<T, +Drop<T>, +Len<T>, +IndexView<T, usize, @u8>> of ByteReader<T> {
+// use core::ops::index::IndexView;
+impl ByteReaderImpl<
+    T, +Drop<T>, +Len<T>, impl IndexViewImpl: IndexView<T, usize>, +Into<IndexViewImpl::Target, @u8>
```
```suggestion
    T, +Drop<T>, +Len<T>, +IndexView<T, usize>, +Into<IndexView::Target, @u8>
```

milancermak commented on packages/storage/src/list.cairo:
```diff
@@ -14,6 +15,9 @@ pub struct List<T> {
 }
 
+#[deprecated(
+    feature: "deprecated-list-trait", note: "Use `starknet::storage::Vec`.", since: "2.7.0"
```
![](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExZXhlYWo2ODgxanVpajBsdGIybGgxN24yejUzYms2d3Y5Nzh5YXhlYyZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/lNA3pbJxd2nmuMcOvH/giphy.gif)

milancermak reviewed (commented):
Random things I've noticed (not necessarily enabled by 2.7.0):
* in `bytes/src/utils.cairo`, `uint_min` can be removed, replaced by `core::cmp::min`
* there's probably something that can be done with iterators on ByteReader, Queue, Stack, Vec and elsewhere(?)


Since we now have const arrays (❤️ `dw` ❤️), IMO it's worth adding optimized `pow2` and `pow10` as these are the most commonly used power functions. Even in Alexandria, all the bitshift fns use `pow(2, n)`.

gaetbout commented on packages/bytes/src/storage.cairo:
```diff
@@ -131,16 +133,16 @@ fn inner_write_bytes(
-                0
-            },
-        };
+        let (tmp, did_overflow) = index_in_chunk.overflowing_add(1);
```
Doesn't work :/

gaetbout commented on packages/math/src/extended_euclidean_algorithm.cairo:
```diff
@@ -31,8 +31,9 @@ pub fn extended_euclidean_algorithm(a: u128, b: u128) -> (u128, u128, u128) {
-    let (bottom, _) = u128_overflowing_mul(quotient, temp);
-    a = u128_wrapping_sub(old_a, bottom);
+    let (bottom, _) = quotient.overflowing_mul(temp);
+    let (a_tmp, _) = old_a.overflowing_sub(bottom);
```
Doesn't work :/

gaetbout commented on packages/linalg/src/kron.cairo:
```diff
@@ -22,14 +22,12 @@ pub fn kron<T, +Mul<T>, +Copy<T>, +Drop<T>,>(
-                }
-        };
+    for x_value in xs {
+        let mut ys_clone = ys;
```
needed otherwise the span is consumed and on the 2nd loop it fails

milancermak replied:
Tests pass however 😱

gaetbout replied:
lemme double check, maybe I messed up somewhere else while testing :D

gaetbout replied:
Actually running in a lowering crash

milancermak replied:
Got the crash as well, submitted an issue to Scarb. Might be worth just removing the `mut` for now?

gaetbout replied:
Done

gaetbout commented:
> Random things I've noticed (not necessarily enabled by 2.7.0):
> 
> * in `bytes/src/utils.cairo`, `uint_min` can be removed, replaced by `core::cmp::min`
> * there's probably something that can be done with iterators on ByteReader, Queue, Stack, Vec and elsewhere(?)
> 
> Since we now have const arrays (❤️ `dw` ❤️), IMO it's worth adding optimized `pow2` and `pow10` as these are the most commonly used power functions. Even in Alexandria, all the bitshift fns use `pow(2, n)`.

updated utils.cairo
y, but I'd leave it for another PR as those are "extra" changes and this PR is big enough

Also would leave for another PR 😅 , please open an issue for that

enitrat commented:
@gaetbout we should suppress the warnings on compilation otherwise it's going to appear on all dependent repositories

gaetbout commented:
I removed all warnings that are from the lib. 
I believe the new warnings are from using now deprecated Alexandria stuff

gaetbout commented on packages/math/src/u512_arithmetics.cairo:
```diff
@@ -1,18 +1,23 @@
+    note: "Use `use core::num::traits::OverflowingAdd`.",
+    since: "2.7.0"
+)]
+pub fn u256_overflow_add(lhs: u256, rhs: u256) -> Result<u256, u256> implicits(RangeCheck) {
```
tbh we could remove all the code that is now deprecated, but to leave some time to the ppl that might be using it, maybe better to add a warning until next version

enitrat replied:
yeah but this has been wrongly been made public in the first place :/

gaetbout replied:
@0xLucqs Opinion?

gaetbout replied:
I updated it

enitrat commented on packages/bytes/src/bytes.cairo:
```diff
@@ -4,8 +4,8 @@ use alexandria_bytes::utils::{
 use alexandria_math::sha256::sha256;
```
should be replaced in favor of syscall

enitrat replied:
here this shouldn't be imported in the `bytes` module. the syscall should be imported here and used for sha256(@Bytes)

gaetbout replied:
solved at 846621a1b731590cf316f9bec20e75cfe39b88a4, lmk if good :)

enitrat commented on packages/bytes/src/bytes.cairo:
```diff
@@ -116,8 +117,18 @@ pub trait BytesTrait {
+    )]
     fn keccak(self: @Bytes) -> u256;
     /// sha256 hash
+    #[deprecated(
```
shouldnt be deprecated. `Bytes` != `ByteArray` - and if we consider this deprecated then the whole Bytes type should be deprecated.

enitrat replied:
`sha256(self: @Bytes)` should use the `sha256` syscall

gaetbout replied:
Anyway since it is using a deprecated fn (math::sha256) the deprecate will propagate upwards.
Maybe we can make this whole bytes stuff deprecated indeed

gaetbout replied:
Should we make the whole Byte deprecated?

enitrat commented on packages/bytes/src/bytes.cairo:
```diff
@@ -116,8 +117,18 @@ pub trait BytesTrait {
     /// concat with other Bytes
     fn concat(ref self: Bytes, other: @Bytes);
     /// keccak hash
+    #[deprecated(
```
same as below

enitrat commented on packages/bytes/src/storage.cairo:
```diff
@@ -86,16 +87,17 @@ fn inner_read_bytes(address_domain: u32, address: StorageAddress) -> SyscallResu
-            },
-        };
+
+        let (tmp, did_overflow) = index_in_chunk.overflowing_add(1);
```
using `checked_add` instead of overflowing_add would solve the problem. same for all subsequent cases

gaetbout replied:
Fixed where it could be done in e8546cf

enitrat commented:
my point is: what currently is used inside alexandria itself should not be triggering warnings

gaetbout commented:
@0xLucqs ?

gaetbout commented:
@0xLucqs ? :D

0xLucqs commented:
I want @milancermak to approve as well

milancermak commented on packages/merkle_tree/src/merkle_tree.cairo:
```diff
@@ -104,18 +106,18 @@ pub impl MerkleTreeImpl<T, +HasherTrait<T>, +Copy<T>, +Drop<T>> of MerkleTreeTra
-                        self.hasher.hash(*proof_element, current_node)
-                    }
-            };
+        while let Option::Some(proof_element) = proof.pop_front() {
```
Can be modernized as `for proof_element in proof`

gaetbout replied:
Done

milancermak commented on packages/merkle_tree/src/merkle_tree.cairo:
```diff
@@ -129,18 +131,18 @@ pub impl MerkleTreeImpl<T, +HasherTrait<T>, +Copy<T>, +Drop<T>> of MerkleTreeTra
-                        self.hasher.hash(*proof_element, leaf)
-                    };
-            };
+        while let Option::Some(proof_element) = proof.pop_front() {
```
Same as above, `for proof_element in proof`

gaetbout replied:
Done

milancermak reviewed (approved):
LGTM! Leaving just a bunch of nits.

If you fancy, there's a bunch of places where the `while-let` can be updated to a `for-in` (e.g. in array_ext.cairo or sha256.cairo, though the latter is deprecated, so meh).

gaetbout commented on packages/numeric/src/trapezoidal_rule.cairo:
```diff
@@ -1,6 +1,6 @@
 use core::array::SpanTrait;
 use core::num::traits::Zero;
-use core::option::OptionTrait;
+use core::ops::{MulAssign, AddAssign};
```
Removed
